### PR TITLE
add `DeprecatedSetting` to ease 7.0 transition of xpack config renames

### LINKED ISSUE #10616: Rename of config params for x-pack in 7.x produces unhelpful error message
In https://github.com/elastic/logstash/pull/10380, we renamed the `url` and `ca` portions of x-pack feature's elasticsearch configs to `hosts` and `certificate_authority` respectively.

While these were soft-deprecated with helpful messaging in 6.7, a user who attempts to upgrade to 7.x without passing through 6.7 (which is unadvised, but will happen) will encounter the following unhelpful error message and exit:

~~~
[2019-03-29T19:00:47,648][FATAL][logstash.runner          ] An unexpected error occurred! {:error=>#<ArgumentError: Setting "xpack.monitoring.elasticsearch.url" hasn't been registered>, :backtrace=>[REDACTED]}
[2019-03-29T19:00:47,732][ERROR][org.logstash.Logstash    ] java.lang.IllegalStateException: Logstash stopped processing because of an error: (SystemExit) exit
~~~

Ideally, we should reintroduce `url` and `ca` to `7.0` and `7.x` as hard-deprecations that still cause the service to fail to start, but with a more helpful error message.

### PR DESCRIPTION
**Pull request by yaauie**
Merge targets: `7.x`, `7.0`

Resolves #10616

### CODE REVIEW

File: x-pack/lib/config_management/extension.rb [hunk 1]
```diff
@@ -34,6 +34,12 @@ def additionals_settings(settings)
         settings.register(LogStash::Setting::NullableString.new("xpack.management.elasticsearch.ssl.keystore.password"))
         settings.register(LogStash::Setting::String.new("xpack.management.elasticsearch.ssl.verification_mode", "certificate", true, ["none", "certificate"]))
         settings.register(LogStash::Setting::Boolean.new("xpack.management.elasticsearch.sniffing", false))
+
+        # These Settings were renamed and deprecated in 6.x timeframe and removed for 7.0; provide guidance to ease transition.
+        settings.register(LogStash::Setting::DeprecatedAndRenamed.new("xpack.management.elasticsearch.url", "xpack.management.elasticsearch.hosts"))
+        settings.register(LogStash::Setting::DeprecatedAndRenamed.new("xpack.management.elasticsearch.ssl.ca", "xpack.management.elasticsearch.ssl.certificate_authority"))
+
+
```
Comment (colinsurprenant) [inline x-pack/lib/config_management/extension.rb]:
```suggestion
```

File: logstash-core/lib/logstash/settings.rb [hunk 2]
```diff
@@ -586,6 +586,34 @@ def validate(value)
         coerce(value)
       end
     end
+
+    ##
+    # Instances of `DeprecatedSetting` can be registered, but will fail with helpful guidance when encountering any
+    # configuration that attempts to explicitly set the value. They should be used in the Major version immediately
+    # following a deprecation to assist users who are porting forward configurations.
+    class DeprecatedSetting < Setting
```
Comment (colinsurprenant) [inline logstash-core/lib/logstash/settings.rb]: 👍

**Comment by colinsurprenant** [comment]
(also relates to #10380)
minor comment - LGTM - thanks @yaauie !

**Comment by yaauie** [comment]
Unless we have another 7.0.0 RC (which is unlikely at this point), this will have missed the boat.

This will be available in `7.0.1` and `7.1.0`.

**Comment by jsvd** [comment]
@yaauie it seems this missed the master branch, and it's only present in 7.x, please "forward" port it unless you see a reason not to have DeprecatedSetting in 8.0

**Pull request closed without merging**